### PR TITLE
[scripts-audit] vcpkg_execute_*

### PR DESCRIPTION
**Pull request by strega-nil**
This goes through the `vcpkg_execute` functions and audits them from issue #17691.

It additionally adds the following:

* `z_vcpkg_forward_output_variable` macro - this allows one to forward to parent scope what makes sense
* `Z_VCPKG_EXECUTE_PROCESS_NAME` - either `_execute_process` or `execute_process`, whichever is allowed to execute in download mode, for use via `cmake_language(CALL)`

### CODE REVIEW

File: scripts/cmake/vcpkg_execute_required_process.cmake [hunk 1]
```diff
@@ -48,72 +48,94 @@ This should be a unique name for different triplets so that the logs don't confl
...
     endif()
+
     # pass output parameters back to caller's scope
-    foreach(arg OUTPUT_VARIABLE ERROR_VARIABLE)
-        if(vcpkg_execute_required_process_${arg})
-            set(${vcpkg_execute_required_process_${arg}} ${${vcpkg_execute_required_process_${arg}}} PARENT_SCOPE)
-        endif()
-    endforeach()
+    if(output_and_error_same)
+        z_vcpkg_forward_output_variable(arg_OUTPUT_VARIABLE out_err_var PARENT_SCOPE)
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_execute_required_process.cmake:135]: Why does this one have PARENT_SCOPE but the others do not?

File: scripts/cmake/vcpkg_execute_required_process.cmake [hunk 2]
```diff
@@ -48,72 +48,94 @@ This should be a unique name for different triplets so that the logs don't confl
...
+        # arg_ERROR_VARIABLE = arg_OUTPUT_VARIABLE, so no need to set it again
+    else()
+        z_vcpkg_forward_output_variable(arg_OUTPUT_VARIABLE out_var)
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_execute_required_process.cmake:138]: Does this do the right thing if one or both of those are not defined?

File: scripts/cmake/vcpkg_execute_required_process_repeat.cmake [hunk 3]
```diff
@@ -15,39 +15,69 @@ vcpkg_execute_required_process_repeat(
 #]===]
 
 function(vcpkg_execute_required_process_repeat)
-    # parse parameters such that semicolons in options arguments to COMMAND don't get erased
-    cmake_parse_arguments(PARSE_ARGV 0 vcpkg_execute_required_process_repeat "ALLOW_IN_DOWNLOAD_MODE" "COUNT;WORKING_DIRECTORY;LOGNAME" "COMMAND")
-    #debug_message("vcpkg_execute_required_process_repeat(${vcpkg_execute_required_process_repeat_COMMAND})")
-    if (DEFINED VCPKG_DOWNLOAD_MODE AND NOT vcpkg_execute_required_process_repeat_ALLOW_IN_DOWNLOAD_MODE)
+    cmake_parse_arguments(PARSE_ARGV 0 arg
+        "ALLOW_IN_DOWNLOAD_MODE"
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_execute_required_process_repeat.cmake:19]: These don't match `usage`

File: scripts/cmake/vcpkg_execute_required_process_repeat.cmake [hunk 4]
```diff
@@ -15,39 +15,69 @@ vcpkg_execute_required_process_repeat(
...
+    if(DEFINED arg_UNPARSED_ARGUMENTS)
+        message(WARNING "${CMAKE_CURRENT_FUNCTION} was passed extra arguments: ${arg_UNPARSED_ARGUMENTS}")
+    endif()
+    foreach(required_arg IN ITEMS COUNT WORKING_DIRECTORY LOGNAME COMMAND)
+        if(NOT DEFINED arg_${required_arg})
+            message(FATAL_ERROR "${required_arg} must be specified.")
+        endif()
+    endforeach()
+
+    if(COUNT LESS "1")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_execute_required_process_repeat.cmake:33]: Does this also check that COUNT is an integer?

File: scripts/cmake/vcpkg_execute_required_process_repeat.cmake [hunk 5]
```diff
@@ -15,39 +15,69 @@ vcpkg_execute_required_process_repeat(
...
+        if(error_code EQUAL "0")
+            set(success TRUE)
             break()
         endif()
-    endforeach(loop_count)
-    if (NOT SUCCESSFUL_EXECUTION)
-        z_vcpkg_prettify_command_line(vcpkg_execute_required_process_repeat_COMMAND_PRETTY ${vcpkg_execute_required_process_repeat_COMMAND})
+    endforeach()
+
+    if(NOT success)
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_execute_required_process_repeat.cmake:65]: Why not test error_code EQUAL "0" here and avoid duplicate `success` state? Alternately, change break() above to return() and then we know it's a failure if we get out of the foreach.

File: scripts/cmake/vcpkg_execute_required_process.cmake (hunk 1 above)
Comment (strega-nil-ms) [reply scripts/cmake/vcpkg_execute_required_process.cmake:135]: A bug.

File: scripts/cmake/vcpkg_execute_required_process_repeat.cmake (hunk 4 above)
Comment (strega-nil-ms) [reply scripts/cmake/vcpkg_execute_required_process_repeat.cmake:33]: yeah

File: scripts/cmake/vcpkg_execute_required_process_repeat.cmake (hunk 5 above)
Comment (ras0219-msft) [reply scripts/cmake/vcpkg_execute_required_process_repeat.cmake:65]: I don't see where `success` is defined -- marking this as unresolved

File: scripts/cmake/vcpkg_execute_in_download_mode.cmake [hunk 6]
```diff
@@ -6,62 +6,60 @@ Execute a process even in download mode.
...
-    endforeach()
+    endif()
+    if(DEFINED arg_RESULT_VARIABLE)
+        set(result_variable_param RESULT_VARIABLE result_var)
+    endif()
+    if(DEFINED arg_RESULTS_VARIABLE)
+        set(results_variable_param RESULTS_VARIABLE results_var)
+    endif()
+
+    cmake_language(CALL "${Z_VCPKG_EXECUTE_PROCESS_NAME}"
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_execute_in_download_mode.cmake:48]: I'd really appreciate a decent-sized comment here documenting specifically a set of test cases that are correctly handled by this construct, focusing on the variety of special characters that CMake supports (`;`, `\`, and combinations thereof)

For example, what happens if I do:

```cmake
vcpkg_execute_in_download_mode(
    COMMAND "a;b" "c" "d\;e" "\\" "f\\;g"
)
```

and how (if at all) does it differ from the equivalent `execute_process()`?

File: scripts/cmake/z_vcpkg_forward_output_variable.cmake [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#[===[.md:
+# z_vcpkg_forward_output_variable
+
+This macro is a simple forward-er to PARENT_SCOPE.
```
Comment (ras0219-msft) [inline scripts/cmake/z_vcpkg_forward_output_variable.cmake:4]:
```suggestion
This macro helps with forwarding values from inner function calls, through a local function scope, into pointer out parameters.
```

File: docs/maintainers/internal/z_vcpkg_forward_output_variable.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# z_vcpkg_forward_output_variable
```
Comment (ras0219-msft) [inline docs/maintainers/internal/z_vcpkg_forward_output_variable.md:1]: It would be good if regenerate.ps1 included obvious text on the page itself that the function is internal in addition to putting it in the `internal/` path (could be a separate PR, but doesn't affect package ABI).

File: scripts/cmake/z_vcpkg_forward_output_variable.cmake [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+```cmake
+z_vcpkg_forward_output_variable(output_var var_to_forward)
```
Comment (ras0219-msft) [inline scripts/cmake/z_vcpkg_forward_output_variable.cmake:7]:
```suggestion
z_vcpkg_forward_output_variable(ptr_to_parent_var local_var)
```

File: scripts/cmake/z_vcpkg_forward_output_variable.cmake [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+is equivalent to
+
+```cmake
+if(DEFINED output_var)
+    if(DEFINED value_var)
+        set("${output_var}" "${value_var}" PARENT_SCOPE)
+    else()
+        unset("${output_var}" PARENT_SCOPE)
+    endif()
+endif()
```
Comment (ras0219-msft) [inline scripts/cmake/z_vcpkg_forward_output_variable.cmake:19]:
```suggestion
if(DEFINED ptr_to_parent_var)
    if(DEFINED local_var)
        set("${ptr_to_parent_var}" "${local_var}" PARENT_SCOPE)
    else()
        unset("${ptr_to_parent_var}" PARENT_SCOPE)
    endif()
endif()
```

File: scripts/cmake/z_vcpkg_forward_output_variable.cmake [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+```
+#]===]
```
Comment (ras0219-msft) [inline scripts/cmake/z_vcpkg_forward_output_variable.cmake:21]:
````suggestion

Take special note that the first argument should be a local variable that has a value of the parent variable name. Most commonly, this local is the result of a pointer-out parameter to a function. If the variable in the first parameter is not defined, this function does nothing, simplifying functions with optional out parameters.

Most commonly, this should be used in cases like:
```cmake
function(my_function out_var)
    file(SHA512 "somefile.txt" local_var)
    z_vcpkg_forward_output_variable(out_var local_var)
endfunction()
```
#]===]
````

File: scripts/cmake/vcpkg_execute_in_download_mode.cmake (hunk 6 above)
Comment (strega-nil-ms) [reply scripts/cmake/vcpkg_execute_in_download_mode.cmake:48]: It doesn't differ at all from the equivalent `execute_process()`.

```cmake
include("scripts/cmake/z_vcpkg_function_arguments.cmake")

function(print_args msg)
	z_vcpkg_function_arguments(args 1)
	message(STATUS "${msg}: ${args}")
endfunction()

function(wrapper)
	z_vcpkg_function_arguments(args)
	message(STATUS "wrapper: ${args}")
	print_args("raw    " ${args})
	cmake_language(CALL print_args "CALL   " ${args})
endfunction()

wrapper("a;b" "c" [[d\;e]] [[\]] [[f\\;g]])
```

prints out

```no-highlight
-- wrapper: a\;b;c;d\\;e;\;f\\\;g
-- raw    : a\;b;c;d\\;e;\;f\\\;g
-- CALL   : a\;b;c;d\\;e;\;f\\\;g
```

you'll notice that the `\` argument gets pushed into the next argument, but this is just... kind of impossible to fix in CMake.
```

**Comment by strega-nil-ms** [comment]
Closed for rollup PR #18838

**Comment by dg0yt** [comment]
Closed but still eating CI for rebuilding the world ...

**Comment by BillyONeal** [comment]
> Closed but still eating CI for rebuilding the world ...

Thanks for the note, I cancelled that run
